src/app/shared/services/chat/offline/chat.flow.ts

```diff
@@ -0,0 +1,207 @@
+                        let latest = flowEvents[flowEvents.length - 1];
+                        if (latest.flowId !== action.flow.uuid) {
+                            flowEvents.push({
+                                flowId: action.flow.uuid,
+                                status: "start"
+                            });
+                            this.flowStatus$.next(flowEvents);
+                        }
+                    }
+                } else {
+                    console.error("Action was to enter_flow however no object for which flow to enter");
+                }
+            }
+            if (action.type === "send_msg" && action.text) {
+                this.doSendMessageAction(action);
+            }
+            if (action.type === "set_contact_field") {
+                this.doSetContactFieldAction(action);
+            }
+        }
+        if (!node.router) {
+            let firstExitWithDestination = node.exits
+                .filter((exit) => exit.destination_uuid)[0];
+            if (firstExitWithDestination) {
+                console.log("Entered node by exiting from node with no router")
+                this.enterNode(this.getNodeById(firstExitWithDestination.destination_uuid));
+            } else {
+                let flowEvents = this.flowStatus$.getValue();
+                flowEvents.push({
+                    flowId: this.flowObject.uuid,
+                    status: "completed"
+                });
+                this.flowStatus$.next(flowEvents);
+            }
+        } else {
+            console.log("Router here?");
+            if (node.router.operand.indexOf("@input.") < 0) {
+                this.useRouter(node);
+            }
+        }
+    }
+
+    private useUserInputRouter(node: RapidProFlowExport.Node, incomingMsg: string) {
+        let matchingCategoryId: string;
+        for (let routerCase of node.router.cases) {
+            let matchesCase: boolean = false;
+            if (routerCase.type === "has_any_word") {
+                matchesCase = this.matchHasAnyWordCase(routerCase, incomingMsg);
+            } else if (routerCase.type === "has_number_between") {
+                matchesCase = this.matchHasNumberBetweenCase(routerCase, incomingMsg);
+            } else if (routerCase.type === "has_number_lt") {
+                matchesCase = this.matchHasNumberLessThanCase(routerCase, incomingMsg);
+            } else if (routerCase.type === "has_number_gt") {
+                matchesCase = this.matchHasNumberGreaterThanCase(routerCase, incomingMsg);
+            }
+            if (matchesCase) {
+                matchingCategoryId = routerCase.category_uuid;
+                break;
+            }
+        }
+        if (matchingCategoryId) {
+            this.exitUsingCategoryId(node, matchingCategoryId);
+        } else {
+            console.warn("Nothing matches :(");
+        }
+    }
+
+    private useRouter(node: RapidProFlowExport.Node) {
```

Comment:
Are routers a rapidpro thing, or like angular router? Maybe a brief comment might be useful to explain what the different userRouter cases do